lib/internal/modules/esm/loader.js

```diff
@@ -408,41 +474,96 @@ class ESMLoader {
 
     const namespaces = await PromiseAll(new SafeArrayIterator(jobs));
 
-    return wasArr ?
-      namespaces :
-      namespaces[0];
+    if (!wasArr) return namespaces[0];
+
+    for (let i = 0; i < count; i++) {
+      const namespace = ObjectCreate(null);
+      namespace.url = specifiers[i];
+      namespace.exports = namespaces[i];
+
+      namespaces[i] = namespace;
+    }
+
+    return namespaces;
   }
 
   /**
    * Provide source that is understood by one of Node's translators.
    *
-   * The internals of this WILL change when chaining is implemented,
-   * depending on the resolution/consensus from #36954
-   * @param {string} url The URL/path of the module to be loaded
+   * @param {URL['href']} url The URL/path of the module to be loaded
    * @param {object} context Metadata about the module
-   * @returns {object}
+   * @returns {{ format: ModuleFormat, source: ModuleSource }}
    */
   async load(url, context = {}) {
-    const defaultLoader = this.#loaders[0];
+    const loaders = this.#loaders;
+    let i = loaders.length - 1;
+    let {
+      fn: loader,
+      url: loaderFilePath,
+    } = loaders[i];
+    let chainFinished = i === 0;
```

Comment:
Seeing `i` here makes me think we’re in a loop or about to declare one, like `while`; but that doesn’t seem to be the case? Somehow `this.#loaders` is getting shorter as each loader is processed? I’m not really following what this code is doing.

Reply:
It's effectively a backwards iterator and `i` is tracking the position within the stack: `i` starts at the top of the hook stack, and each time `next()` is called, `i` steps towards the bottom. The stack itself is not actually changed, just to which entry in the stack `i` is pointing.

loaderHooks: `A`, `B`, `C`, `D`

initially → `loaderHook = D`
`next()` → `loaderHook = C`
`next()` → `loaderHook = B`
`next()` → `loaderHook = A`

When `i` reaches `0`, it's done (`chainFinished` is `true`).

Reply:
I took another look into abstracting this. I think I was right before: It seems like a good idea but actually is way too much trouble/complex.

I added some code docs to both methods to explain how/why it works.

Reply:
You probably beat me to it, but I was going to suggest adding some comments and renaming `i` to something else, like maybe `hookIndex`.

Reply:
I didn't rename `i`, but I can do if you'd still prefer (it's used exactly like a while loop's `i`).